### PR TITLE
nrf54h20: add exmif peripheral DTS description and sample configuration

### PR DESCRIPTION
**Pull request by e-rk**

### CODE REVIEW

File: samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&uart136 {
+	status = "okay";
+};
```
Comment (gmarull) [inline samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:9]: why is this needed?

File: samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+&mx25uw63 {
+	status = "okay";
+};
+
+&gpio6 {
+	status = "okay";
+};
```
Comment (gmarull) [inline samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:25]: same: where is this used in this overlay?

File: boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts [hunk 3]
```diff
@@ -195,3 +195,27 @@ ipc0: &cpuapp_cpurad_ipc {
...
 	hw-flow-control;
 };
+
+&exmif {
+	cs-gpios = <&gpio6 3 GPIO_ACTIVE_LOW>;
+	pinctrl-0 = <&exmif_default>;
+	pinctrl-names = "default";
+	mx25uw63: mx25uw6345g@0 {
+		compatible = "jedec,spi-nor";
+		status = "disabled";
```
Comment (gmarull) [inline boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts:205]: why is it disabled at board level?

File: include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h [hunk 4]
```diff
@@ -126,6 +126,28 @@
...
+/** EXMIF DQ5 */
+#define NRF_FUN_EXMIF_DQ5 41U
+/** EXMIF DQ6 */
+#define NRF_FUN_EXMIF_DQ6 42U
+/** EXMIF DQ7 */
+#define NRF_FUN_EXMIF_DQ7 43U
+/** EXMIF CS0 */
+#define NRF_FUN_EXMIF_CS0 44U
+/** EXMIF CS1 */
+#define NRF_FUN_EXMIF_CS1 45U
```
Comment (gmarull) [inline include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h:150]: content in the wrong commit?

**Review by gmarull — DISMISSED** [review]

File: include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h (hunk 4 above)
Comment (e-rk) [reply include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h:150]: This is the first commit that makes use of those definitions, so I put them here. Should those definitions be in a separate commit?

File: boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts (hunk 3 above)
Comment (e-rk) [reply boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts:205]: I can enable it by default, if that's the recommendation. On nRF54H20 DK the pins used by EXMIF/external flash are not connected to anything else.

File: samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay (hunk 1 above)
Comment (e-rk) [reply samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:9]: I think it is a leftover. I see that `uart136` is enabled by default now.

Comment (e-rk) [reply samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay]: Removed

File: boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts (hunk 3 above)
Comment (gmarull) [reply boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts:205]: at board level things are usually enabled by default.

File: include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h (hunk 4 above)
Comment (gmarull) [reply include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h:150]: How does the pinctrl driver configure exmif?

Comment (e-rk) [reply include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h:150]: On nRF54H20 it's not needed. The configuration is done by nrf-regtool.

File: boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts (hunk 3 above)
Comment (e-rk) [reply boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts:210]: Enabled EXMIF by default, but flash node is kept disabled, like it was done here: https://github.com/zephyrproject-rtos/zephyr/blame/main/boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuapp.dts#L135

File: dts/common/nordic/nrf54h20.dtsi [hunk 5]
```diff
@@ -987,6 +987,19 @@
 				status = "disabled";
 				interrupts = <470 NRF_DEFAULT_IRQ_PRIORITY>;
 			};
+
+			exmif: spi@95000 {
```
Comment (anangl) [inline dts/common/nordic/nrf54h20.dtsi:991]: Please keep the nodes ordered by address and locate this node right after the `usbhs` one.

File: dts/common/nordic/nrf54h20.dtsi [hunk 6]
```diff
@@ -987,6 +987,19 @@
...
+				compatible = "nordic,nrf-exmif", "snps,designware-spi";
```
Comment (anangl) [inline dts/common/nordic/nrf54h20.dtsi:992]: I don't think that "snps,designware-spi" should be listed here. It is not actually possible to use this peripheral that way without Nordic specific stuff.

File: include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h (hunk 4 above)
Comment (anangl) [reply include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h:150]: I think it would be better to add this definitions in a separate commit, with a note in its message explaining why no changes in `pinctrl_nrf.c` are needed.

File: boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts (hunk 3 above)
Comment (gmarull) [reply boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts:210]: why?

File: include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h (hunk 4 above)
Comment (gmarull) [reply include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h:150]: > I think it would be better to add this definitions in a separate commit, with a note in its message explaining why no changes in `pinctrl_nrf.c` are needed.

agreed

File: boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts (hunk 3 above)
Comment (e-rk) [reply boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts:210]: I wanted to avoid having `CONFIG_SPI_NOR` enabled in all samples by default. Setting status okay would result in `CONFIG_SPI_NOR` being enabled by default when `CONFIG_FLASH=y`. On current nRF54H20 DKs, powering up the flash chip requires some additional steps from the user.

File: dts/common/nordic/nrf54h20.dtsi [hunk 7]
```diff
@@ -987,6 +987,19 @@
...
+				#address-cells = <1>;
+				#size-cells = <0>;
+				reg = <0x95000 0x500 0x95500 0xb00>;
+				reg-names = "wrapper", "core";
+				interrupts = <149 NRF_DEFAULT_IRQ_PRIORITY>;
+				clock-frequency = <DT_FREQ_M(400)>;
```
Comment (gmarull) [inline dts/common/nordic/nrf54h20.dtsi:998]: what is the clock source here? looks like this should be referencing a clock?

File: dts/common/nordic/nrf54h20.dtsi (hunk 6 above)
Comment (e-rk) [reply dts/common/nordic/nrf54h20.dtsi:992]: Removing the compatible results in spi_dw.c not being compiled and `DT_INST_FOREACH_STATUS_OKAY` not picking up the node.

File: dts/common/nordic/nrf54h20.dtsi (hunk 7 above)
Comment (e-rk) [reply dts/common/nordic/nrf54h20.dtsi:998]: The clock source frequency is fixed and the application domain will only be able to turn it on or off. Currently there is no service that allows turning it on or off, but it is enabled by default.

Comment (gmarull) [reply dts/common/nordic/nrf54h20.dtsi:998]: I mean that from a DT perspective, looks like there should be a clock that runs at 400MHz defined somewhere, unless this peripheral has a contained 400MHz clock

Comment (gmarull) [reply dts/common/nordic/nrf54h20.dtsi:998]: image

File: dts/common/nordic/nrf54h20.dtsi (hunk 5 above)
Comment (e-rk) [reply dts/common/nordic/nrf54h20.dtsi]: Changed the order

File: include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h (hunk 4 above)
Comment (e-rk) [reply include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h:150]: Split the change to a separate commit.

**Comment by e-rk** [comment]
I added one more commit which modified the nRF54H20 Kconfig.defconfig. It is needed to select the correct register layout for the SPI DW peripheral.

File: boards/nordic/nrf54h20dk/Kconfig.defconfig [hunk 8]
```diff
@@ -15,3 +15,6 @@ config BT_CTLR
 	default y if BT
 
 endif # BOARD_NRF54H20DK_NRF54H20_CPURAD
+
+config SPI_DW_HSSI
+	default y if SPI_DW
```
Comment (gmarull) [inline boards/nordic/nrf54h20dk/Kconfig.defconfig:20]: is this really a board-level defconfig?

Comment (e-rk) [reply boards/nordic/nrf54h20dk/Kconfig.defconfig]: Right, that was the wrong one. Moved to soc defconfig.

File: dts/common/nordic/nrf54h20.dtsi (hunk 7 above)
Comment (gmarull) [reply dts/common/nordic/nrf54h20.dtsi]: so it looks like you should define a `fixed-clock` from an application perspective (similar to fll16m), as it's manages by sysctrl, then reference it as `clocks = <&hsfll_exmif>;`

image

Comment (e-rk) [reply dts/common/nordic/nrf54h20.dtsi]: Added the clock source.

File: dts/common/nordic/nrf54h20.dtsi [hunk 9]
```diff
@@ -93,6 +93,12 @@
 			#clock-cells = <0>;
 			clock-frequency = <DT_FREQ_M(16)>;
 		};
+
+		exmif_hsfll: exmif_hsfll {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <DT_FREQ_M(400)>;
+		};
```
Comment (gmarull) [inline dts/common/nordic/nrf54h20.dtsi:101]: note: commit message says app needs to turn this on/off, so this is not really a fixed-clock, it will require a clock control driver that uses nrfs underneath. Looks like we'll need a new compatible, eg `nordic,nrf-hsfll-remote` or similar and a driver so we can use `clock_control_on/off`. @anangl any comments here?

Comment (e-rk) [reply dts/common/nordic/nrf54h20.dtsi:101]: Should this be a part of this PR, or could this be addressed later?

Comment (anangl) [reply dts/common/nordic/nrf54h20.dtsi:101]: AFAICT the driver only supports the fixed clock frequency now: https://github.com/zephyrproject-rtos/zephyr/blob/0b87cb1545d5ec99b4f5e8368bb04dfbee4098b1/drivers/spi/spi_dw.c#L636-L641 so I think there is no point in trying to provide proper handling of that HSFLL here. I think the initial approach with the `clock-frequency` property in the `exmif` node was fine. Especially taking into account that the main idea behind this PR is to just take advantage of the already existing `spi_dw` driver to use the EXMIF peripheral in its basic SPI mode, temporarily, until we have a proper EXMIF driver.

Comment (e-rk) [reply dts/common/nordic/nrf54h20.dtsi:101]: Looking at the above snippet the previous approach with `clock-frequency` gets the job done just the same as `clocks`. I'll stay with the current approach (`clocks`) unless some problem emerges.

Comment (anangl) [reply dts/common/nordic/nrf54h20.dtsi:101]: Yes, it does the job in the `spi_dw` driver context, but the HSFLL is not actually a fixed clock, as its output frequency can be changed, so the added node is a bit misleading.

Comment (e-rk) [reply dts/common/nordic/nrf54h20.dtsi]: Okay, I restored it to the previous approach.

File: samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay [hunk 10]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&mx25uw63 {
```
Comment (anangl) [inline samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:7]: The order of commits is wrong. The above node label is added in the third commit but it is referred in the first one.

Comment (anangl) [reply samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:7]: nit: also "soc: nrf54h: Enable SPI DW HSSI register layout" should be before "boards: nordic: add mx25uw6345g flash node to nRF54H20dk" that activates the spi_dw driver.

**Review by anangl — APPROVED** [review]

File: samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay (hunk 10 above)
Comment (e-rk) [reply samples/drivers/jesd216/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:7]: Done

**Review by carlescufi — APPROVED** [review]
